rfd/0035-aws-node-join.md

```diff
@@ -0,0 +1,294 @@
+  - AWS will check the signature and return an error if it is invalid, in which case the node will be rejected.
+3. Extract the Account from the `sts:GetCallerIdentity` response, if the request was provided.
+4. Extract the Organization from the `organizations:DescribeOrganization` response, if the request was provided.
+5. Check the Organization and Account against the configured deny rules. Reject the node if any match.
+6. Check the Organization and Account against the configured allow rules. Reject the node if none match.
+7. Possibly extra checks to prevent replay attacks, see the following section.
+
+By presenting signed AWS API requests, the node proves that it has access to
+AWS credentials. We rely on AWS to verify the signatures.
+
+The API responses prove the Organization and Account of the credentials
+used to sign the requests. We check that the request is a real AWS API
+endpoint, and rely on AWS to return the real identity for the signed
+requests.
+
+Nodes never share their AWS credentials, only signed requests.
+
+### Mitigating the Risk of Replay Attacks
+
+Signed API requests will only be held in memory and never logged or written to
+disk. All transport of signed API requests will be over TLS.
+
+We can include arbitrary headers in the signed API request and AWS will verify
+the signature over those headers. We can use this to include a timestamp on the
+request and enforce a TTL. Amazon does this by default with a TTL of 15
+minutes.  We could set a shorter TTL ourselves, or make it configurable.
+
+We could also include a UUID header, and use this to verify that each signed
+request is only used once by having the auth server store the UUID of requests
+that have been used recently and still have a valid TTL. With multiple auth
+servers this would require some backend coordination.
+
+### Teleport Configuration
+
+Support for new configuration options will be added to Teleport's configuration
+file in order to enable and configure this feature.
+
+```yaml
+teleport:
+
+  # This section should be used on nodes which will join the cluster with the
+  # new aws join method, in place of auth_token.
+  aws_token:
+    # credential_source is the AWS credentials source. The main supported value
+    # will be "ec2", where the node will automatically fetch aws credentials
+    # from the EC2 instance metadata. Support for a "file" or "env" type may be
+    # added, which would be especially useful for testing.
+    credential_source:
+      type: ec2
+
+    # claims configures which signed requests the node should create and send
+    # to the auth server. The two supported values are:
+    # - "identity" for the sts:GetCallerIdentity request
+    # - "organization" for the organizations:DescribeOrganization request
+    # We could default to just send both, but I think this makes more sense if
+    # the node does not have organization permissions, or if we wish to add
+    # more claims in the future.
```

Comment:
Which ones are required by default? do we need organization claim at all times? If we need both to prove the identity I would remove the configuration parameter for now too

Reply:
The idea was to let the admin choose if they want to authenticate with the organization, the account, or both. Some users may just let any node in the organization join and will not create any account-based rules, in which case the `sts:GetCallerIdentity` request is not necessary. Same if they just want to go by account, then the `organizations:DescribeOrganization` request is not necessary.

But it's probably simpler to remove the configuration for now and just send both

Reply:
For the initial rollout, the less configuration options and permutations, the better